<a href="https://colab.research.google.com/github/chrislee973/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

In [1]:
### YOUR WORK HERE 

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data


--2020-09-07 18:37:56--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data’

imports-85.data     100%[===================>]  25.33K  --.-KB/s    in 0.1s    

2020-09-07 18:37:57 (194 KB/s) - ‘imports-85.data’ saved [25936/25936]



In [6]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.names

--2020-09-07 18:39:48--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.names
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4747 (4.6K) [application/x-httpd-php]
Saving to: ‘imports-85.names’

imports-85.names    100%[===================>]   4.64K  --.-KB/s    in 0s      

2020-09-07 18:39:49 (96.2 MB/s) - ‘imports-85.names’ saved [4747/4747]



In [9]:
#Read in dataset

column_names = ['symboling', 'normalized-losses', 'make', 'fuel-type', 
                'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 
                'engine-location', 'wheel-base', 'length', 'width', 'height', 
                'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 
                'fuel-system', 'bore', 'stroke', 'compression-ratio', 
                'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

df = pd.read_csv('imports-85.data', header=None, names = column_names )
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [10]:
#Replace '?' with np.NaN

df = df.replace('?', np.NaN)
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [13]:
#Obtain random sample
df_sample = df.sample(10, random_state=30)
df_sample

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
103          0               108         nissan  ...       19          25  13499
26           1               148          dodge  ...       31          38   7609
20           0                81      chevrolet  ...       38          43   6575
155          0                91         toyota  ...       27          32   8778
185          2                94     volkswagen  ...       27          34   8195

[10 rows x 26 columns]

In [16]:
#Highway-mpg mean for entire dataset 

df_highway_mean = df['highway-mpg'].mean()
df_highway_mean

30.75121951219512

In [24]:
#One sample t-test

stats.ttest_1samp(df_sample['highway-mpg'], df_highway_mean, nan_policy='omit')

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

**Using a confidence interval of 95%, we get a t-value of .89 and a pvalue of .396. Thus, we can say that the sample mean we obtained is not significantly different from the population mean.**


The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [22]:
#One sample t-test

stats.ttest_1samp(df_sample['highway-mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

**The null hypothesis is that the cars he sell has a highway-mpg of 35. 
The alternative hypothesis is that the cars he sell don't have a highway-mpg of 35.** 

**Using a confidence interval of 95%, we get a t-statistic of .95 and pvalue of .36. Thus, we fail to reject the null hypothesis based on our sample of 10 cars.** 

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [20]:
df_sample_100 = df.sample(100, random_state=30)
df_sample_100

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
..         ...               ...            ...  ...      ...         ...    ...
189          3               NaN     volkswagen  ...       24          29  11595
192          0               NaN     volkswagen  ...       33          38  13845
171          2               134         toyota  ...       24          30  11549
203         -1                95          volvo  ...       26          27  22470
47           0               145         jaguar  ...       15          19  32250

[100 rows x 26 columns]

In [23]:
#T-test

stats.ttest_1samp(df_sample_100['highway-mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

**With a t-statistic of -6 and a pvalue of 3.0e-09, we have enough data/evidence to reject the null hypothesis that the cars he sells typically has a highway-mpg of 35.**

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

**We drew different conclusions about the salesman's claim because the different sizes of the samples we drew gave us either, in the case of the first sample, insufficient data to say anything statistically significant about the claim, or in the case of second sample, sufficient data to draw a statistically significant conclusion about the claim.** 

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [25]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-09-07 19:35:12--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-09-07 19:35:13 (137 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [26]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers)

print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [27]:
df = df.replace({'y': 1, 'n': 0, "?": np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [29]:
#Split df into two df's based on political party 

repub = df[df['party']== 'republican']


dem = df[df['party']=='democrat']


party  handicapped-infants  ...  duty-free  south-africa
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0
5    democrat                  0.0  ...        1.0           1.0
6    democrat                  0.0  ...        1.0           1.0
..        ...                  ...  ...        ...           ...
425  democrat                  0.0  ...        1.0           NaN
426  democrat                  1.0  ...        1.0           1.0
428  democrat                  NaN  ...        1.0           1.0
429  democrat                  1.0  ...        1.0           1.0
431  democrat                  0.0  ...        0.0           1.0

[267 rows x 17 columns]

In [46]:
repub['physician-fee-freeze'].mean()

0.9878787878787879

In [45]:
dem['physician-fee-freeze'].mean()

0.05405405405405406

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

**Null hypothesis**: Democratic level of support for the 'synfueles' bill is equal to Republican level of support for the 'synfuels' bill.

**Alternative Hypothesis:**  Democratic level of support for the 'synfueles' bill is **not** equal to Republican level of support for the 'synfuels' bill.

In [32]:
stats.ttest_ind(repub['synfuels'], dem['synfuels'], nan_policy='omit')

Ttest_indResult(statistic=-8.293603989407588, pvalue=1.5759322301054064e-15)

**Based on the pvalue of 1.5e-15, we are confident on the 99% level in rejecting the null hypothesis that dem and repub level of support for the synfuels bill is equal. We recommend the alternative hypothesis that the levels of support for the bill between the two parties are different. The Democrat level of support for the synfuels bill is far greater than the republican level of support, as evidenced by the large negative t-statistic.**

**Null hypothesis**: Democratic level of support for the 'religious_groups' bill is equal to Republican level of support for the 'synfuels' bill.

**Alternative Hypothesis:**  Democratic level of support for the 'religious_groups' bill is **not** equal to Republican level of support for the 'synfuels' bill.

In [47]:
stats.ttest_ind(repub['education'], dem['education'], nan_policy='omit')

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

**Based on the pvalue of 2.4e-20, we are confident on the 99% level in rejecting the null hypothesis that dem and repub levels of support for the religious_groups bill are equal. We suggest the alternative hypothes that the levels of support for the bill between the two parties are different. The Republican level of support for the religious_groups bill is far greater than the democratic level of support, as evidenced by the large positive t-statistic.**

**Null hypothesis**: Democratic level of support for the 'water-project' bill is equal to Republican level of support for the 'water-project' bill.

**Alternative Hypothesis:**  Democratic level of support for the 'water-project' bill is **not** equal to Republican level of support for the 'water-project' bill.



In [53]:
stats.ttest_ind(repub['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

**Based on the pvalue of .929, and with a confidence interval of 90%, we fail to reject the null hypothesis. That is, differences in support level between Democrats and Republicans for the water project bill is not significant at the 90% level.**

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [52]:

for bill in df.columns: 
  if bill != 'party': 
    t_test = stats.ttest_ind(repub[bill], dem[bill], nan_policy='omit')
    print(f"The t-test for the {bill} bill is:\n {t_test}\n")


The t-test for the handicapped-infants bill is:
 Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

The t-test for the water-project bill is:
 Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

The t-test for the budget bill is:
 Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

The t-test for the physician-fee-freeze bill is:
 Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

The t-test for the el-salvador-aid bill is:
 Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)

The t-test for the religious-groups bill is:
 Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)

The t-test for the anti-satellite-ban bill is:
 Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31)

The t-test for the aid-to-contras bill is:
 Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)

The t-test for the mx-mi



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE